src/Build/Graph/ProjectGraph.cs

```diff
@@ -330,44 +396,33 @@ private enum NodeState
         }
 
         /// <remarks>
-        /// Load a graph with root node at entryProjectFile
+        /// Traverse an evaluated graph
         /// Maintain the state of each node (InProcess and Processed) to detect cycles
         /// returns false if loading the graph is not successful
         /// </remarks>
-        private (bool success, List<string> projectsInCycle) ProcessNode(ConfigurationMetadata projectToEvaluate,
-            Dictionary<ConfigurationMetadata, NodeState> nodeState,
+        private (bool success, List<string> projectsInCycle) DetectCycles(ProjectGraphNode node,
+            Dictionary<ProjectGraphNode, NodeState> nodeState,
             ProjectCollection projectCollection,
             PropertyDictionary<ProjectPropertyInstance> globalProperties)
         {
-            nodeState[projectToEvaluate] = NodeState.InProcess;
-            ProjectGraphNode parsedProject = CreateNewNode(projectToEvaluate, projectCollection);
-            IEnumerable<ProjectItemInstance> projectReferenceItems = parsedProject.ProjectInstance.GetItems(ProjectReferenceItemName);
+            nodeState[node] = NodeState.InProcess;
+            IEnumerable<ProjectItemInstance> projectReferenceItems = node.ProjectInstance.GetItems(ProjectReferenceItemName);
             foreach (var projectReferenceToParse in projectReferenceItems)
             {
-                if (!string.IsNullOrEmpty(projectReferenceToParse.GetMetadataValue(ToolsVersionMetadataName)))
-                {
-                    throw new InvalidOperationException(string.Format(
-                        CultureInfo.InvariantCulture,
-                        ResourceUtilities.GetResourceString(
-                            "ProjectGraphDoesNotSupportProjectReferenceWithToolset"),
-                        projectReferenceToParse.EvaluatedInclude,
-                        parsedProject.ProjectInstance.FullPath));
-                }
-
                 string projectReferenceFullPath = projectReferenceToParse.GetMetadataValue(FullPathMetadataName);
```

Comment:
Is this item/metadata manipulation required? Since this is now pushed to after graph construction, can you use the graph directly?

Reply:
the project references are not set up till this point even though all graph nodes have been constructed.

Reply:
This seems unfortunate. How hard would it be to fix?

Reply:
One option to fix this is to construct a ProjectGraphNode object even before evaluation completes. This way, there would be no initialization done in the constructor of each node.
The other option would be to keep a list of parent nodes in tasksInProgress. Once a task completes, add project references to each of them.